gonat_linux.go

```diff
@@ -127,15 +128,52 @@ func (s *server) dispatch() {
 	for {
 		select {
 		case pkt := <-s.fromDownstream:
+			log.Tracef("Got packet from downstream: %v", pkt.FT())
 			s.onPacketFromDownstream(pkt)
 		case pkt := <-s.fromUpstream:
+			log.Tracef("Got packet from upstream: %v", pkt.FT())
 			s.onPacketFromUpstream(pkt)
+		case c := <-s.closingConns:
+			s.deleteAndCloseConn(c)
 		case c := <-s.closedConns:
-			s.deleteConntrackEntry(c.upFT)
+			s.finalizeConn(c)
 		case <-reapTicker.C:
 			s.reapIdleConns()
 		case <-s.close:
-			return
+			if s.numConns == 0 {
+				// All connections already closed
+				return
+			}
+
+			// Close all connections
+			for _, c := range s.connsByDownFT {
+				c.Close()
```

Comment:
Won't this block? `c.Close()` waits for the server to read off `s.closingConns`

Reply:
It used to, but it doesn't anymore

Reply:
IIUC, it ends up calling `s.requestCloseConn`, which will block until something reads off `s.closingConns`.  This will not happen though because the dispatch loop is waiting for `c.Close` to return.  Basically a deadlock where the dispatch loop is waiting for `c.Close` and `c.Close` is waiting for the dispatch loop.

Reply:
Just calling `c.Close` in a new goroutine would be an easy fix.  Maybe we'd spawn a lot of goroutines, but only momentarily.

Reply:
Good point! `s.closingConns` is a buffered channel, so for writes to it to block, it would have to be full. It's possible that this happens sometimes, which would explain why we only see this problem sometimes!